google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/sdk/ManagedCloudSdkUpdateService.java

```diff
@@ -0,0 +1,140 @@
+ */
+
+package com.google.cloud.tools.intellij.appengine.sdk;
+
+import com.google.cloud.tools.intellij.GctFeature;
+import com.google.cloud.tools.intellij.service.PluginInfoService;
+import com.google.common.annotations.VisibleForTesting;
+import com.intellij.notification.Notification;
+import com.intellij.openapi.components.ServiceManager;
+import java.awt.event.ActionListener;
+import java.time.Clock;
+import java.util.Timer;
+import java.util.TimerTask;
+
+/** Schedules and manages automatic updates for {@link ManagedCloudSdkService}. */
+public class ManagedCloudSdkUpdateService {
+  // one week.
+  @VisibleForTesting static final long SDK_UPDATE_INTERVAL_MS = 1000 * 60 * 60 * 24 * 7;
+  // 20 seconds to show notification and then proceed with update.
+  private static final int UPDATE_NOTIFICATION_DELAY_MS = 1000 * 20;
+
+  @VisibleForTesting
+  static final String SDK_UPDATER_THREAD_NAME = "managed-google-cloud-sdk-updates";
+
+  static ManagedCloudSdkUpdateService getInstance() {
+    return ServiceManager.getService(ManagedCloudSdkUpdateService.class);
+  }
+
+  private Timer updateTimer;
+  private TimerTask sdkUpdateTask;
+
+  void activate() {
+    if (!CloudSdkServiceUserSettings.getInstance().isAutomaticUpdateEnabled()) {
+      return;
+    }
+
+    if (updateTimer == null) {
+      updateTimer = new Timer(SDK_UPDATER_THREAD_NAME);
+    }
+    // cancel tasks from previous activation.
+    if (sdkUpdateTask != null) {
+      sdkUpdateTask.cancel();
+    }
+
+    sdkUpdateTask =
+        new TimerTask() {
+          @Override
+          public void run() {
+            doUpdate();
+          }
+        };
+    long lastTimeOfUpdate =
+        CloudSdkServiceUserSettings.getInstance().getLastAutomaticUpdateTimestamp();
+    long delayBeforeCheckMillis = SDK_UPDATE_INTERVAL_MS;
+    long now = getClock().millis();
+    if (lastTimeOfUpdate != 0 && lastTimeOfUpdate < now) {
+      delayBeforeCheckMillis =
+          Math.min(
+              SDK_UPDATE_INTERVAL_MS,
+              Math.max(0, SDK_UPDATE_INTERVAL_MS - (getClock().millis() - lastTimeOfUpdate)));
+    }
+
+    schedule(sdkUpdateTask, delayBeforeCheckMillis, SDK_UPDATE_INTERVAL_MS);
+  }
+
+  /** Called when managed SDK update operation (update or install) successfully completes. */
+  void notifySdkUpdate() {
+    CloudSdkServiceUserSettings.getInstance().setLastAutomaticUpdateTimestamp(getClock().millis());
+  }
+
+  @VisibleForTesting
+  Clock getClock() {
+    return Clock.systemDefaultZone();
+  }
+
+  @VisibleForTesting
+  void schedule(TimerTask timerTask, long delay, long period) {
+    updateTimer.scheduleAtFixedRate(timerTask, delay, period);
+  }
+
+  @VisibleForTesting
+  javax.swing.Timer createUiTimer(int delayMillis) {
+    return new javax.swing.Timer(delayMillis, null);
+  }
+
+  private void doUpdate() {
+    CloudSdkService cloudSdkService = CloudSdkService.getInstance();
+    if (cloudSdkService instanceof ManagedCloudSdkService) {
```

Comment:
do you need the `instanceof` here? can' t you do `CloudSdkService#isInstallSupported()` instead?

Reply:
oh I see, you are casting it below. I guess you add an #isUpToDate() method to the interface too, but I guess thats not semantically correct (since you have to return `true` or `false` always for the custom case)..

Reply:
Exactly, adding too many methods would make custom SDK implementation look bloated and return strange values. Update always work with Managed SDK so I guess this makes it OK to be strongly coupled.